pkg/controller/egressgateway/egressgateway_controller.go

```diff
@@ -0,0 +1,666 @@
+	}
+
+	return nil
+}
+
+// Blank assignment to verify that ReconcileEgressGateway implements reconcile.Reconciler.
+var _ reconcile.Reconciler = &ReconcileEgressGateway{}
+
+// ReconcileEgressGatewayLayer reconciles a EgressGatewayLayer object.
+type ReconcileEgressGateway struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver.
+	client          client.Client
+	scheme          *runtime.Scheme
+	provider        operatorv1.Provider
+	status          status.StatusManager
+	clusterDomain   string
+	licenseAPIReady *utils.ReadyFlag
+	usePSP          bool
+}
+
+// Reconcile reads that state of the cluster for an EgressGateway object and makes changes
+// based on the state read and what is in the EgressGateway.Spec.
+func (r *ReconcileEgressGateway) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling EgressGateway")
+
+	// Get all the Egress Gateway resources available.
+	egws, err := getEgressGateways(ctx, r.client)
+	if err != nil {
+		reqLogger.Error(err, "Error querying for Egress Gateway")
+		r.status.SetDegraded(operatorv1.ResourceReadError, "Error querying for Egress Gateway", err, reqLogger)
+		return reconcile.Result{}, err
+	}
+
+	// If there are no Egress Gateway resources, return.
+	if len(egws) == 0 {
+		r.status.OnCRNotFound()
+		return reconcile.Result{}, nil
+	}
+
+	/* Reconcile is done as follows.
+	1. At the start, assume all the EGW resources need to be reconciled.
+	2. If the request is to a particular EGW resource, find the requested EGW from the list of EGWs.
+	3. If the requested EGW resource is not present, it could have been deleted.
+	   'egws' is now the list of all EGW resources present. Get the cumulative status and update
+	   TigeraStatus object.
+	4. If the requested EGW resource is present, then it is the only resource to reconcile. Remove this
+	   EGW resource from the list and get the status of all the other EGW resources present and update
+	   TigeraStatus accordingly.
+	5. If the request is not to a particular EGW resource, reconcile all the resources.
+	*/
+
+	// egwsToReconcile is the list of Egress Gateway resources that needs to be reconciled.
+	// To start with all EGW resources must be reconciled.
+	egwsToReconcile := egws
+	if request.Namespace != "" {
+		requestedEGW, idx := getRequestedEgressGateway(egws, request)
+		if requestedEGW == nil {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			reqLogger.Info("EgressGateway object not found")
+			// Since the EGW resource is not found, remove the deployment.
+			r.status.RemoveDeployments(types.NamespacedName{Name: request.Name, Namespace: request.Namespace})
```

Comment:
I think this type of controller might warrant the addition of a `SetDeployements` function that allows us to add / remove all at once, rather than try to add / remove the right deployments in the controller. I think it will simplify the logic and make it less likely we accidentally forget to add or remove a deployment in a certain branch. 

We can include this in a separate PR, though.